hbase-server/src/main/java/org/apache/hadoop/hbase/master/procedure/SnapshotProcedure.java

```diff
@@ -0,0 +1,422 @@
+    if (env.getMasterCoprocessorHost() != null) {
+      env.getMasterCoprocessorHost()
+        .postCompletedSnapshotAction(ProtobufUtil.createSnapshotDesc(snapshot), htd);
+    }
+    status.markComplete("Snapshot " + snapshot.getName() + "  completed");
+  }
+
+  private void snapshotOfflineRegions(MasterProcedureEnv env) throws IOException {
+    List<RegionInfo> regions = env.getAssignmentManager()
+      .getTableRegions(snapshotTable, false).stream()
+      .filter(r -> RegionReplicaUtil.isDefaultReplica(r))
+      .filter(RegionInfo::isSplit).collect(Collectors.toList());
+
+    ThreadPoolExecutor exec = SnapshotManifest
+      .createExecutor(env.getMasterConfiguration(), "OfflineRegionsSnapshotPool");
+    try {
+      ModifyRegionUtils.editRegions(exec, regions, new ModifyRegionUtils.RegionEditTask() {
+        @Override
+        public void editRegion(final RegionInfo regionInfo) throws IOException {
+          snapshotManifest.addRegion(CommonFSUtils.getTableDir(rootDir, snapshotTable), regionInfo);
+        }
+      });
+    } finally {
+      exec.shutdown();
+    }
+    status.setStatus("Completed referencing offline regions of table: " + snapshotTable);
+  }
+
+  private void snapshotMobRegion() throws IOException {
+    RegionInfo mobRegionInfo = MobUtils.getMobRegionInfo(htd.getTableName());
+    snapshotManifest.addMobRegion(mobRegionInfo);
+    status.setStatus("Completed referencing HFiles for the mob region of table: " + snapshotTable);
+  }
+
+  @Override
+  protected void serializeStateData(ProcedureStateSerializer serializer) throws IOException {
+    super.serializeStateData(serializer);
+    serializer.serialize(SnapshotProcedureStateData
+      .newBuilder().setSnapshot(this.snapshot).build());
+  }
+
+  @Override
+  protected void deserializeStateData(ProcedureStateSerializer serializer) throws IOException {
+    super.deserializeStateData(serializer);
+    SnapshotProcedureStateData data = serializer.deserialize(SnapshotProcedureStateData.class);
+    this.snapshot = data.getSnapshot();
+  }
+
+  private Procedure<MasterProcedureEnv>[] createRemoteSnapshotProcedures(MasterProcedureEnv env) {
+    return env.getAssignmentManager().getTableRegions(snapshotTable, true)
+      .stream().filter(r -> RegionReplicaUtil.isDefaultReplica(r))
+      .map(r -> new SnapshotRegionProcedure(snapshot, r))
+      .toArray(SnapshotRegionProcedure[]::new);
+  }
+
+  // here we assign region snapshot manifest verify tasks to all region servers
+  // in cluster with the help of master load balancer.
+  private Procedure<MasterProcedureEnv>[] createRemoteVerifyProcedures(MasterProcedureEnv env)
+    throws IOException {
+    List<RegionInfo> regions = env
+      .getAssignmentManager().getTableRegions(snapshotTable, false);
+    List<ServerName> servers = env
+      .getMasterServices().getServerManager().getOnlineServersList();
+    return env.getMasterServices().getLoadBalancer()
```

Comment:
Do we need to use LoadBalancer here? I suppose we could use something like the SplitWALManager, where we could configure the max concurrency number for a region server, and if there are more tasks then the number_of_region_servers * max_concurrency, we just do not schedule all the tasks, once there are tasks finished, we will schedule new tasks. This could make the tasks more evenly distributed.

Reply:
That's a better idea to distribute the tasks. I'll address it ASAP.